src/Features/CSharp/Portable/Completion/KeywordRecommenders/DataKeywordRecommender.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.CSharp.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.KeywordRecommenders
+{
+    internal class DataKeywordRecommender : AbstractSyntacticSingleKeywordRecommender
+    {
+        public DataKeywordRecommender()
+            : base(SyntaxKind.DataKeyword)
+        {
+        }
+
+        protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
+        {
+            return
+                context.IsGlobalStatementContext ||
+                context.TargetToken.IsUsingKeywordInUsingDirective() ||
+                IsValidContextForType(context, cancellationToken);
+        }
+
+        private static bool IsValidContextForType(CSharpSyntaxContext context, CancellationToken cancellationToken)
+        {
+            if (context.IsTypeDeclarationContext(
+                validModifiers: SyntaxKindSet.AllTypeModifiers,
+                validTypeDeclarations: SyntaxKindSet.ClassInterfaceStructTypeDeclarations,
+                canBePartial: true,
+                cancellationToken: cancellationToken))
+            {
+                return CheckPreviousModifiers(context);
+            }
+
+            return false;
```

Comment:
do we have test with things like `ref data partial` struct (and permutations)?

Reply:
question still applies

Reply:
Sorry, I thought I addressed that comment. We have a test for `partial $$` and one for `ref $$`. I don't think that what follows the $$ matters, but I can add such tests if you think it's useful.

Reply:
`data` is a contextual keyword only directly before `class` at the moment, so if anything here relies on `data` being a contextual keyword, that won't work.

Reply:
@agocke I've already accounted for that, since you won't have typed `class` at the time you want to type `data`. We had a similar problem with `partial` or `ref` I think.